scripts/algorithms/fm.dml

```diff
@@ -0,0 +1,114 @@
+# X         String   ---      Location to read the input matrix X to be factorized
+# y         String   ---      Location to read the input matrix y corresponding to the feature vector
+# k         Int      ---      hyperparameter
+# reg       String   "L2"	    Regularization: 
+#                             "L2" = L2 regularization;
+#                             "wL2" = weighted L2 regularization
+# lambda_w  Double   0.000001 Regularization parameter, no regularization if 0.0
+# lambda_v  Double   0.000001
+# y_est     String   ---      Output file for estimated target values
+# error 
+# -------------------------------------------------------------------------------------------------------
+# Example: An extended design matrix with four real valued predictors. 
+# -------------------------------------------------------------------------------------------------------
+#  A  B  C | TI NH SW ST |TI   NH   SW   ST | t | TI NH SW ST |   y   | x(i)
+# ----------------------------------------------------------- --------  ---- 
+#  1  0  0   1  0  0  0   0.3  0.3  0.3  0.0  13  0  0  0  0  |   5   | x(1) # feature vector x(1) & y(=5) its corresponding target
+#  1  0  0   0  1  0  0   0.3  0.3  0.3  0.0  14  1  0  0  0  |   3   | x(2)
+#  1  0  0   0  0  1  0   0.3  0.3  0.3  0.0  16  0  1  0  0  |   1   | x(3)
+#  0  1  0   0  0  1  0   0.0  0.0  0.5  0.5   5  0  0  0  0  |   4   | x(4)
+#  0  1  0   0  0  0  1   0.0  0.0  0.5  0.5   8  0  0  1  0  |   5   | x(5)
+#  0  0  1   1  0  0  0   0.5  0.0  0.5  0.0   9  0  0  0  0  |   1   | x(6)
+#  0  0  1   0  0  1  0   0.5  0.0  0.5  0.0  12  1  0  0  0  |   5   | x(7)
+# ----------------------------------------------------------- --------  ---- ----------------------------
+#  user    | movies      | other movies the |time| Last movie |target |
+#  (active)| (active)    | user has rated   |    | rated      |rating |  
+# --------------------------------------------------------------------  ---- ----------------------------
+
+fileX = $X; # nxm 
+fileY = $y; # nx1
+fileY_est = $y_est; # nx1
+
+# Default values for some parameters
+lambda_w  = ifdef ($lambda_w, 0.000001);
+lambda_v  = ifdef ($lambda_v, 0.000001);
+fileO     = ifdef ($O, " ");
+fileLog   = ifdef ($Log, " ");
+fmtY_est  = ifdef ($fmt, "text");
+
+
+X = read (fileX);
+y = read (fileY);
+
+w0 = matrix(0, rows=1, cols=1); # 1x1
+W  = matrix(0, rows=1, cols=n); # 1xn
+V  = rand(rows=n, cols=k, min=0.0, max=1.0); # nxk
+
+loss = 0;
+
+# model equation 
+y_est = w0 + sum(W * X) + 0.5 * sum( ( X %*% t(V) ) - ( (X) ^ 2 ) %*% t( V ^ 2)); # whether to take (X != 0) instead of X or not
+
+# norm_L2 = ( lambda_w * (W ^ 2) + lambda_v * (V ^ 2) ); 
+# error = (y - y_est)^2;
+# loss = error + norm_L2;
+
+# ----------------------------------------------------------------------------------------------------------------
+# adam model | Jimmy Lei Ba & Diederik P. Kingma | see for more details: (arXiv:1412.6980v9 [cs.LG] 30 Jan 2017) | 
```

Comment:
The Adam optimizer is already implemented as a DML function in [`nn/optim/adam.dml`](https://github.com/apache/systemml/blob/master/scripts/nn/optim/adam.dml).  Can you please use that function here?  You can simply add `source("nn/optim/adam.dml") as adam` near the top of the file and the engine will find the import.  Use `adam::init(...)` for each parameter matrix of interest to initialize the associated 1st & 2nd moment vectors, and `adam::update(...)` inside the main training loop for each parameter matrix of interest to apply updates.  There is an example that uses the SGD w/ Nesterov momentum optimizer (nn/optim/sgd_nesterov.dml) that follows the same pattern as the Adam optimizer in the [MNIST LeNet example](https://github.com/apache/systemml/blob/master/scripts/nn/examples/mnist_lenet.dml).